File: 15. Two Pointers - 3Sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */

/*
  I: array of numbers
  O: array of triplets, representing numbers that sum to 0
  C:
  E:

  Strategy:
    Options:
      1. Recursion (i, j, k)
      2. Two (three) pointers (loop through each, then drop number farthest from zero)
*/
var threeSumOne = function(nums) {

  let solutions = []

  // Sort nums first
  nums = nums.sort((a, b) => (a-b))

  let checked = {}
  let left = 0
  let right = nums.length - 1
  let center = 1

  while ((right - left) >= 2) {

    // Check if sum === 0
    if (nums[left] + nums[center] + nums[right] === 0) {

      // check if we've already evaluated this one
      let array = [nums[left], nums[center], nums[right]].sort()
      if (!checked[array.join('')]) {
        solutions.push([nums[left], nums[center], nums[right]])
      }
      checked[array.join('')] = true
    }

    center++
    // When we've reached the end for one iteration, then reset pointers
    if (center >= right) {
      // move pointer which is farthest from zero to speed up iterations
      right--
      center = left + 1
    }



  }

  return solutions
};

let threeSumTwo = (nums) => {

  let solutions = {}
  nums = nums.sort((a,b) => (a-b))

  let left = 0
  let center = 0
  let right = nums.length - 1

  while (right - left >= 2) {
    center++

    if (center === right) {

      if ((nums[left] + nums[right]) < 0) {
        left++
      } else {
        right--
      }
      center = left
      continue;
    }

    let sum = nums[left] + nums[center] + nums[right]
    if (sum === 0) {
      solutions[`${nums[left]},${nums[center]},${nums[right]}`] = true
    }

  }

  return Object.keys(solutions).map(val => val.split(','))

}

let threeSum = (nums) => {
  nums = nums.sort((a,b) => (a-b))
  let solutions = {}

  // Step through every number
  for (i = 0; i < nums.length; i++) {

    // i = left
    let center = i + 1
    let right = nums.length - 1

    while (center < right) {
      let sum = nums[i] + nums[center] + nums[right]

      if (sum > 0) {
        right--
      } else if (sum < 0) {
        center++
      } else { // we have a match
        // get rid of duplicates with a sorted array and saving that string
        let array = [nums[i],nums[center],nums[right]].sort()
        solutions[array.join(',')] = true
        center++
      }
    }
  }

  // split our solution object back out into an array
  return Object.keys(solutions).map(val => val.split(','))
}

// === TESTING ===
console.log(threeSum([-1,0,1,2,-1,-4]))
console.log(threeSum([0, 0, 0]))
console.log(threeSum([0,0,0,0]))
console.log(threeSum([-1,0,1,0]))
console.log(threeSum([3,0,-2,-1,1,2]))
console.log(threeSum([-1,0,1,2,-1,-4,-2,-3,3,0,4]))

File: 15. Two Pointers - 3Sum.py
'''
  Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.

  Notice that the solution set must not contain duplicate triplets

  I: array of integers
  O: array of arrays (int combinations that add up to zero
  C: must not contain duplicates
  E:

  Strategy:
    - Sort list
    - Two (three) pointers from left and right, and middle
    - Maintain solutions in a set so we don't have duplicates
'''

class Solution:
    def threeSum(self, nums):
        solutions = {} # key: solution (hyphen-concatenated), val: True (always)
        returnArray = []
        total = 0

        # Sort original list of nums
        nums.sort()

        # Two (three) pointers (indices)
        left = 0
        mid = 1
        right = len(nums) - 1

        # Loop until left and right meet
        while (left < right - 1):
            mid = left + 1

            while (mid < right):
                # Calculate new total each time
                total = nums[left] + nums[mid] + nums[right]

                # If total == 0, then add to solution
                if (total == 0):
                    s = f'{nums[left]}:{nums[mid]}:{nums[right]}'

                    # Check if we've already saved solution
                    try:
                        solutions[s]
                    except:
                        solutions[s] = True
                        returnArray.append([nums[left], nums[mid], nums[right]])
                elif (total > 0): # getting farther from solution, so just adjust left and right
                    break


                mid += 1

            # Check which is farther from zero, and adjust accordingly
            if nums[left] + nums[right] < 0:
                left += 1
            else:
                right -= 1

        return returnArray


# ==== TESTING ====
# print(Solution.threeSum(None, [-1,0,1,2,-1,-4]))
# print(Solution.threeSum(None, [0, 0, 0, 0]))
print(Solution.threeSum(None, [-1,0,1,2,-1,-4,-2,-3,3,0,4]))